splicemachine/spliceengine
# DB-11795 Improve index lookup cost estimation

## Short Description
This change improves index lookup cost estimation process.

## Long Description
In investigating query hash `SMopSK5W`, three problems are identified (for detailed explanation, see DB-11718). This PR addresses the second issue.

Currently, our index lookup cost has a penalty of 4000 per row. There are two issues about this. Firstly, the penalty is too high compared to actual execution times. Secondly, it's an over simplification because index lookup happens in a batched manner and could have some parallelism. Batch size and the maximum number of batches that can run concurrently are controlled by settings in `SQLConfiguration`. This change tunes the index lookup cost to consider these parameters and lower lookup penalty.

Concrete coefficients and separate formulae for OLTP and OLAP are based on the index lookup benchmark results obtained from DB-11737.

## How to test
After this change, index lookup cost should be proportional to index lookup execution time with roughly a fixed coefficient. Specifically, on OLTP we should see index lookup costs also have the "flat" phase when there are still capacity for running index lookup batches concurrently, and suddenly bumped up when one more serial stage needed. On OLAP, we should see that index lookup cost is roughly linear but with a high startup cost, making looking up small amount of rows less attractive.

To test index lookup costs really share the same trend as the execution times obtained from running the microbenchmark, expected costs are extracted from the microbenchmark results and we assert the estimated costs are close enough. Currently, without tuning anything else (scan costs, parallelism, etc.), we see that estimated costs fall into 11% range of expected costs on OLTP, and 14% range on OLAP for all test cases.

See ITs in `IndexLookupCostEstimationIT` for details.
